### PR TITLE
[Firebreak] Runnable in production

### PR DESCRIPTION
**Pull request by 46bit**
What
----

This PR contains a sequence of commits that took this repo from the output of a SPIKE into something I was able to run in production.

How to review
-----

I suggest reviewing the codebase that results from this, but the commit messages should be quite explanatory too.

I already merged https://github.com/alphagov/paas-auditor/pull/2 to make this PR fewer than a million lines, so take a look at that too

Who can review
-----

Not @46bit

### CODE REVIEW

**Comment by richardTowers** [comment]
Comments on the code on master before I review this PR:

* The README.md and version are still from paas-billing
* There aren't any tests yet
* I'm not a big fan of having paas-billing's git history in this repo, but I'm happy to let that slide
* Otherwise looks fairly uncontroversial - not too much hangover from paas-billing

File: Makefile [hunk 1]
```diff
@@ -12,8 +12,8 @@ run-dev: bin/paas-auditor run-dev-exports
 run-dev-exports:
 	## Runs the application with local credentials
 	$(eval export CF_API_ADDRESS=${CF_API_ADDRESS})
-	$(eval export CF_CLIENT_ID=paas-auditor)
-	$(eval export CF_CLIENT_SECRET=$(shell aws s3 cp s3://gds-paas-${DEPLOY_ENV}-state/cf-vars-store.yml - | awk '/uaa_clients_paas_auditor_secret/ { print $$2 }'))
+	$(eval export CF_CLIENT_ID=paas-billing)
+	$(eval export CF_CLIENT_SECRET=$(shell aws s3 cp s3://gds-paas-${DEPLOY_ENV}-state/cf-vars-store.yml - | awk '/uaa_clients_paas_billing_secret/ { print $$2 }'))
```
Comment (richardTowers) [inline Makefile]: Should be `paas-auditor`?

File: db/store.go [hunk 2]
```diff
@@ -178,6 +178,40 @@ func (s *EventStore) GetCfAuditEvents(filter RawEventFilter) ([]cfclient.Event,
 	return events, nil
 }
 
+func (s *EventStore) LastSeenEvent() (*time.Time, error) {
+	ctx, cancel := context.WithTimeout(s.ctx, DefaultQueryTimeout)
+	defer cancel()
+	tx, err := s.db.BeginTx(ctx, &sql.TxOptions{ReadOnly: true})
```
Comment (richardTowers) [inline db/store.go]: Do we actually need a transaction here if we're just doing one select from one table?

**Comment by 46bit** [comment]
> they

thank you :)

>

File: eventcollector/cf_audit_event_collector.go [hunk 3]
```diff
@@ -63,38 +60,65 @@ func (c *CfAuditEventCollector) Run(ctx context.Context) error {
...
 			}
 			elapsedTime := time.Since(startTime)
-			c.eventsCollected += len(collectedEvents)
+			c.eventsCollected += collectedEventsCount
 			c.logger.Info("collected", lager.Data{
-				"count":    len(collectedEvents),
+				"count":    collectedEventsCount,
 				"duration": elapsedTime.String(),
 			})
+			// FIXME: If the database connection has broken, errors swallowed abvoe
```
Comment (richardTowers) [inline eventcollector/cf_audit_event_collector.go]:
```suggestion
			// FIXME: If the database connection has broken, errors swallowed above
```

**Comment by 46bit** [comment]
I've opened some Issues and removed the various `FIXME` comments we had put in. 

@richardTowers Thanks for reviewing up til now. Hopefully this code is now ready for final review. 

As we're not putting this in our pipelines, I'll deploy it in London this afternoon and check it actually still works.

**Comment by tlwr** [comment]
Reviewed the last 2 commits since last approval :shipit:

**Comment by 46bit** [comment]
I've got it running happily in `prod-lon`. Merging.

**Pull request merged**